src/cuda_block_manager.cc

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "cuda_block_manager.h"
+
+#include "cuda_utils.h"
+
+
+namespace triton { namespace core {
+std::unique_ptr<CudaBlockManager> CudaBlockManager::instance_;
```

Comment:
Is this fine as a global? Or should it belong to server object or similar?

Reply:
This is just to make the static instance work. I think it is still created per server object.

Reply:
The singleton would only be created once per program - so even if there were multiple server objects - you would have only one instance of the block manager.

I found this interesting singleton implementation using static within the create function itself: https://stackoverflow.com/a/1008289 

Either is fine - maybe for completeness I would explicitly initialzie insteance_ = nullptr.

If we need a seperate instance per server object - then we will need to create and store the instance in the server object .

Reply:
Yeah, I just followed what we have in the cuda_memory_manager. We follow the pattern mentioned in stack overflow in Python backend for singleton: https://github.com/triton-inference-server/python_backend/blob/889585ce0fc4d4c88eeb69dddbe0f26d08af4b6f/src/pb_utils.h#L262-L266